kani-compiler/src/codegen_cprover_gotoc/reachability.rs

```diff
@@ -0,0 +1,586 @@
+// Copyright Kani Contributors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+//! This module implements a cross-crate collector that allow us to find all items that
+//! should be included in order to verify one or more proof harness.
+//!
+//! This module works as following:
+//!   - Traverse all reachable items starting at the given starting points.
+//!   - For every function, traverse its body and collect the following:
+//!     - Constants / Static objects.
+//!     - Functions that are called or have their address taken.
+//!     - VTable methods for types that are coerced as unsized types.
+//!   - For every static, collect initializer and drop functions.
+//!
+//! We have kept this module agnostic of any Kani code in case we can contribute this back to rustc.
+use rustc_data_structures::fx::FxHashSet;
+use rustc_middle::mir::interpret::{AllocId, ConstValue, ErrorHandled, GlobalAlloc, Scalar};
+use rustc_middle::mir::mono::MonoItem;
+use rustc_middle::mir::visit::Visitor as MirVisitor;
+use rustc_middle::mir::{
+    Body, CastKind, Constant, ConstantKind, Location, Rvalue, Terminator, TerminatorKind,
+};
+use rustc_middle::span_bug;
+use rustc_middle::ty::adjustment::PointerCast;
+use rustc_middle::ty::layout::{
+    HasParamEnv, HasTyCtxt, LayoutError, LayoutOf, LayoutOfHelpers, TyAndLayout,
+};
+use rustc_middle::ty::{
+    Closure, ClosureKind, Const, ConstKind, Instance, InstanceDef, ParamEnv, Ty, TyCtxt, TyKind,
+    TypeAndMut, TypeFoldable, VtblEntry,
+};
+use rustc_span::def_id::DefId;
+use rustc_span::source_map::Span;
+use rustc_target::abi::{HasDataLayout, TargetDataLayout, VariantIdx};
+use tracing::{debug, debug_span, trace, warn};
+
+struct MonoItemsCollector<'tcx> {
+    /// The compiler context.
+    tcx: TyCtxt<'tcx>,
+    /// Set of collected items used to avoid entering recursion loops.
+    collected: FxHashSet<MonoItem<'tcx>>,
+    /// Keep items collected sorted by visiting order which should be more stable.
```

Comment:
Good idea!

Reply:
I'll check if `def_id` will be stable enough. My main concern was the id changing as we modify the compiler.